Correctly propagate write failures from ChunkedWriteHandler

kachayev opened the pull request:
Motivation:

`ChunkedWriteHandler` should report write operation as failed in case *any* chunked was not written. Right now this is not true for the last chunk.

Modifications:

* Check if the appropriate write operation was successful when reporting the last chunk
* Skip writing chunks if the write operation was already marked as "done"

Result:

Fixes https://github.com/netty/netty/issues/8700.

kachayev commented:
Sure!

kachayev commented:
@normanmaurer Actually... From what I can tell, there are no test cases to cover input close and failure propagations at all. I think it makes more sense to cover all of them (including `discard` of pending writes in the queue). Should I add them to this PR or maybe open a separate one?

normanmaurer commented:
Just add them here.... thanks 

> Am 14.01.2019 um 13:19 schrieb Oleksii Kachaiev <notifications@github.com>:
> 
> @normanmaurer Actually... From what I can tell, there are no test cases to cover input close and failure propagations at all. I think it makes more sense to cover all of them (including discard of pending writes in the queue). Should I add them to this PR or maybe open a separate one?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

kachayev commented:
Deal!

normanmaurer commented on handler/src/main/java/io/netty/handler/stream/ChunkedWriteHandler.java:
```diff
@@ -209,6 +209,18 @@ private void doFlush(final ChannelHandlerContext ctx) {
+                // and report end of input, but this doesn't work with any
+                // source wrapped in HttpChunkedInput.
+                // See https://github.com/netty/netty/issues/8700.
+                this.currentWrite = null;
```
do we also need to release the message ?

kachayev replied:
Not here... I assume someone who resolved the promise (either with success or with failure) have already done this by calling `closeInput`. As we're not allocating any new chunks (calling `readChunk`), everything else should be the responsibility of a specific input implementation.

normanmaurer replied:
good point. Maybe add a comment ?

kachayev replied:
Will do!

normanmaurer commented:
@kachayev please ping me once I should review again

kachayev commented:
@normanmaurer Sure, I will

kachayev commented:
@normanmaurer Added 12 new test cases.

normanmaurer commented:
@kachayev <3

normanmaurer commented:
@netty-bot test this please

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -227,6 +228,207 @@ public long progress() {
+                if (++this.passedWrites < 4) {
+                    ctx.write(msg, promise);
+                } else {
+                    promise.tryFailure(new RuntimeException());
```
You also need to call `ReferenceCountUtil.release(msg)` before failing the promise.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -227,6 +228,207 @@ public long progress() {
+
+        EmbeddedChannel ch = new EmbeddedChannel(failLast, new ChunkedWriteHandler());
+        ChannelFuture r = ch.writeAndFlush(new ChunkedFile(TMP, 1024 * 16)); // 4 chunks
+        ch.close().awaitUninterruptibly();
```
just use `ch.finish()` and assert the return value.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -227,6 +228,207 @@ public long progress() {
+        ChannelFuture r1 = ch.write(new ChunkedFile(TMP));
+        ChannelFuture r2 = ch.write(new ChunkedNioFile(TMP));
+        ch.write(notifiableInput);
+        ch.close().awaitUninterruptibly();
```
just use `ch.finish()` and assert the return value.

kachayev replied:
@normanmaurer `ch.finish` would return `false` because we effectively don't do any writes here. All messages are still in `ChunkedWriteHandler` queue, until `flush`. Meaning both `inboundMessages` and `outboundMessages` queues are empty.

normanmaurer replied:
@kachayev sure which is fine just use `assertFalse(...)` as this is what you expect, or what I am missing ?

kachayev replied:
@normanmaurer Oh, yeah. You're right! 🤦‍♂️It's just mentally hard to read `assertFalse(ch.finish())` as "all good, just no messages" :) I'll leave a comment why this is expected.

kachayev replied:
@normanmaurer Update once again. I also "fixed" a couple of older tests to be consistent in terms of `ch.finish()` usage.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -227,6 +228,207 @@ public long progress() {
+        ChannelOutboundHandlerAdapter noOpWrites = new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                promise.tryFailure(new RuntimeException());
```
You also need to call `ReferenceCountUtil.release(msg)` before failing the promise.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -227,6 +228,207 @@ public long progress() {
+
+        EmbeddedChannel ch = new EmbeddedChannel(noOpWrites, new ChunkedWriteHandler());
+        ch.writeAndFlush(nonClosableInput).awaitUninterruptibly();
+        ch.close();
```
just use `ch.finish()` and assert the return value.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -255,4 +457,64 @@ private static void check(Object... inputs) {
+        ChannelOutboundHandlerAdapter noOpWrites = new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                promise.tryFailure(new RuntimeException());
```
Call `ReferenceCountUtil.release(msg)` before failing the promise.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -255,4 +457,64 @@ private static void check(Object... inputs) {
+
+        EmbeddedChannel ch = new EmbeddedChannel(noOpWrites, new ChunkedWriteHandler());
+        ChannelFuture r = ch.writeAndFlush(input).awaitUninterruptibly();
+        ch.close();
```
just use `ch.finish()` and assert the return value.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -255,4 +457,64 @@ private static void check(Object... inputs) {
+                if (alreadyFailed) {
+                    ctx.write(msg, promise);
+                } else {
+                    promise.tryFailure(new RuntimeException());
```
call `ReferenceCountUtil.release(msg)` before failing the promise.

normanmaurer commented on handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java:
```diff
@@ -255,4 +457,64 @@ private static void check(Object... inputs) {
+        EmbeddedChannel ch = new EmbeddedChannel(failFirst, new ChunkedWriteHandler());
+        ChannelFuture r1 = ch.write(input1);
+        ChannelFuture r2 = ch.writeAndFlush(input2).awaitUninterruptibly();
+        ch.close();
```
just use `ch.finish()` and assert the return value.

normanmaurer commented:
@kachayev after my comments are addressed this should pass to build on the CI as it currently fails because of some buffer leaks.

kachayev commented:
@normanmaurer I've added `release` calls. Sorry, missed those in the first place. I also switched to `ch.finish()` where reasonable. There are still a few tests where we don't do outbound writes at all because of `ChunkedWriteHandler` internal machinery (commented them out in the code).

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@kachayev thanks a lot... merged
